src/bottom_up_iterator.jl

```diff
@@ -12,235 +12,212 @@ Concrete implementations of this iterator should implement the following methods
 abstract type BottomUpIterator <: ProgramIterator end
 
 Base.@doc """
-    @programiterator BasicIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
+    @programiterator DepthIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
```

Comment:
The `DepthIterator`is tightly coupled with `BottomUp` iterator, right? It needs to have access to the bank and it will not straight-forwardly work with other iterators? Maybe than we should rename it somehow more informative, like `BUDepthIterator`?